packages/compat/src/default-pipeline.ts

```diff
@@ -1,50 +1,57 @@
 import { App, Addons as CompatAddons, Options, PrebuiltAddons } from '.';
 import { toBroccoliPlugin, PackagerConstructor, Variant, EmberAppInstance } from '@embroider/core';
+import { tmpdir } from '@embroider/shared-internals';
 import { Node } from 'broccoli-node-api';
 import writeFile from 'broccoli-file-creator';
 import mergeTrees from 'broccoli-merge-trees';
+import { createHash } from 'crypto';
+import { join, dirname } from 'path';
+import { sync as pkgUpSync } from 'pkg-up';
 
 export interface PipelineOptions<PackagerOptions> extends Options {
   packagerOptions?: PackagerOptions;
   onOutputPath?: (outputPath: string) => void;
   variants?: Variant[];
 }
 
+export function stableWorkspaceDir(appRoot: string) {
+  let hash = createHash('md5');
+  hash.update(dirname(pkgUpSync({ cwd: appRoot })!));
+  return join(tmpdir, 'embroider', hash.digest('hex').slice(0, 6));
+}
+
 export default function defaultPipeline<PackagerOptions>(
   emberApp: EmberAppInstance,
   packager?: PackagerConstructor<PackagerOptions>,
-  options?: PipelineOptions<PackagerOptions>
+  options: PipelineOptions<PackagerOptions> = {}
 ): Node {
-  let outputPath: string;
   let addons;
+
   if (process.env.REUSE_WORKSPACE) {
     addons = new PrebuiltAddons(emberApp, options, process.env.REUSE_WORKSPACE);
   } else {
     if (process.env.SAVE_WORKSPACE) {
-      if (!options) {
-        options = {};
-      }
       options.workspaceDir = process.env.SAVE_WORKSPACE;
+    } else {
+      options.workspaceDir = stableWorkspaceDir(emberApp.project.root);
     }
+
+    emberApp.project.ui.write(`Building into ${options.workspaceDir}\n`);
     addons = new CompatAddons(emberApp, options);
-    addons.ready().then(result => {
-      if (options && options.onOutputPath) {
-        options.onOutputPath(result.outputPath);
-      } else {
-        console.log(`Building into ${result.outputPath}`);
-      }
-      outputPath = result.outputPath;
-    });
+
+    if (options && options.onOutputPath) {
+      options.onOutputPath(options.workspaceDir);
+    }
```

Comment:
Let's remove this hook for now (instead of changing the semantics fundamental), if we need to bring it back we can rationalize + document the semantics.